### PR TITLE
Closes Forgot Password Flow

### PR DESCRIPTION
**Pull request by dhmi19**
## Notion ticket link

[Forgot Password Flow](https://www.notion.so/uwblueprintexecs/Forgot-Password-Flow-41991540c20940c082d4189080294cdc)

NOTE: This PR doesn't add the shapes to the forgot password page. We should make another ticket to add those graphics to all the main pages as needed. 

## Implementation description
* Added a new page called ForgotPassword.tsx
* Modified SetPassword.tsx page to work for both forgot password and create new account
* Modified authService::sendPasswordSetupLink() function to work for both forgot password and new account flow

## Steps to test
1. Ensure you have an account created for our website and that local storage is clear
2. Go to the login page and click forgot password
3. Enter email and click button -> a confirmation modal should be sent
4. Go to your email and follow the steps from the link
5. You should be able to log in with your new password after going through the verification code and set password steps.

## What should reviewers focus on?
* I used the existing create account and password setup infrastructure for the forgot password flow
* I refactored existing functions to make them flexible for both use cases
* Pay attention to the new backend functions I created/ used -> if there is a better way of doing it, I would be happy to discuss alternatives.

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have run docker-compose up and my project compiled
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

**Comment by dhmi19** [comment]
Note: Still need to add the graphics and do all the design work on the front end. Currently, only the functionality is complete.

File: backend/typescript/rest/authRoutes.ts [hunk 1]
```diff
@@ -103,6 +103,33 @@ authRouter.post("/register", registerRequestValidator, async (req, res) => {
   }
 });
 
+/**
+ * Sends the user an email to reset their password. Used when the user has forgotten their password.
+ * @returns bool - true if successfully sent the email, false otherwise
+ */
+authRouter.post("/forgotPassword", async (req, res) => {
```
Comment (dhmi19) [inline backend/typescript/rest/authRoutes.ts:110]: - [ ] Add documentation for this in swagger.yaml

File: frontend/src/assets/Auth modal graphic.svg [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<svg width="461" height="229" viewBox="0 0 461 229" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (dhmi19) [inline frontend/src/assets/Auth modal graphic.svg:1]: Unstage this from this PR. We need either a graphic the same size as the modal, or a better quality graphic which we custom position in the modal

File: frontend/src/components/auth/ForgotPasswordEmailModal.tsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+const ForgotPasswordEmailModal = ({
+  isOpen,
+  onClose,
+}: {
+  isOpen: boolean;
+  onClose: () => void;
+}): React.ReactElement => {
+  return (
+    <Modal
```
Comment (dhmi19) [inline frontend/src/components/auth/ForgotPasswordEmailModal.tsx:22]: Need to add background graphic here

File: frontend/src/components/auth/Login.tsx [hunk 4]
```diff
@@ -90,6 +92,12 @@ const Login = (): React.ReactElement => {
                 placeholder="Password"
                 errorMessage="The username or password you entered is incorrect."
               />
+              <Link
```
Comment (dhmi19) [inline frontend/src/components/auth/Login.tsx:95]: Add margin top here- but how? :(

File: frontend/src/theme/index.tsx [hunk 5]
```diff
@@ -4,7 +4,7 @@ import "@fontsource/dm-sans";
 
 import { extendTheme } from "@chakra-ui/react";
 
-import PasswordInputFieldStyle from "./PasswordInputFieldStyle";
+import PasswordInputFieldStyle from "./Components/PasswordInputFieldStyle";
```
Comment (dhmi19) [inline frontend/src/theme/index.tsx:7]: Note: We should organize all our custom component styles inside a separate folder so they don't get messy. Thoughts?

File: backend/typescript/services/implementations/authService.ts [hunk 6]
```diff
@@ -182,21 +184,26 @@ class AuthService implements IAuthService {
     try {
       const authId = await this.userService.getAuthIdById(user.id);
 
-      const setPasswordLink = `${process.env.CLIENT_URL}/auth/action?mode=verify-user&uid=${authId}`;
+      const setPasswordLink = `${process.env.CLIENT_URL}/auth/action?mode=verify-user&uid=${authId}&new-account=${isNewAccount}`;
```
Comment (susheelk) [inline backend/typescript/services/implementations/authService.ts:187]: Is there any way to not have this parameter, and have the frontend determine whether this is a new account? This seems like a security risk. 
Also, we should probably discuss sending some random token instead of the authID to improve security, but that might be a separate PR

File: backend/typescript/rest/authRoutes.ts [hunk 7]
```diff
@@ -103,6 +103,33 @@ authRouter.post("/register", registerRequestValidator, async (req, res) => {
...
+  try {
+    // Find the user record with that email. If none found, error will be caught
+    const firebaseUserUid = await authService.getFirebaseUserIdByEmail(
+      req.body.email,
+    );
+
+    // Set the user's password to the random access code. This is needed on the front end to verify the user
+    const accessCode = await authService.setTemporaryUserPassword(
```
Comment (susheelk) [inline backend/typescript/rest/authRoutes.ts:118]: Do we want to change a user's password if they forget it? ie consider the case where they click forgot password, but immediately remember their old password

File: backend/typescript/rest/authRoutes.ts [hunk 8]
```diff
@@ -103,6 +103,33 @@ authRouter.post("/register", registerRequestValidator, async (req, res) => {
...
+    );
+
+    const currentUser = await userService.getUserByEmail(req.body.email);
+
+    // Send email to user to reset their password
+    await authService.sendPasswordSetupLink(currentUser, accessCode, false);
+
+    res.status(200).json();
+  } catch (error: unknown) {
+    sendErrorResponse(error, res);
```
Comment (susheelk) [inline backend/typescript/rest/authRoutes.ts:129]: If the email doesn't exist/is incorrect, this should still probably send a success for security/obfuscation. ie we shouldn't reveal whether an email has a valid account

File: frontend/src/App.tsx [hunk 9]
```diff
@@ -93,6 +99,11 @@ const App = (): React.ReactElement => {
                   path={Routes.AUTH_ACTIONS}
                   component={AuthActions}
                 />
+                <Route
+                  exact={false}
```
Comment (susheelk) [inline frontend/src/App.tsx:103]: What's the different between the two routes you added? Shouldn't we always use exact matching?

File: frontend/src/components/auth/ForgotPassword.tsx [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+          <FormControl mt="1rem">
+            <Box mt="4%" mb="10%">
+              <FormLabel>Email address</FormLabel>
+              <Input
+                value={email}
+                name="email"
+                isInvalid={isEmailInvalid}
+                placeholder="Email address"
+                onChange={(event) => {
+                  setIsEmailInvalid(false);
```
Comment (susheelk) [inline frontend/src/components/auth/ForgotPassword.tsx:95]: Do you validate the email here?

**Review by susheelk — CHANGES_REQUESTED** [review]

File: frontend/src/components/auth/ForgotPassword.tsx (hunk 10 above)
Comment (dhmi19) [reply frontend/src/components/auth/ForgotPassword.tsx:95]: the `onSendEmailClick` function just checks that the email is not null. I thought checking the email format would be unnecessary

File: backend/typescript/rest/authRoutes.ts [hunk 11]
```diff
@@ -103,6 +103,33 @@ authRouter.post("/register", registerRequestValidator, async (req, res) => {
   }
 });
 
+/**
+ * Sends the user an email to reset their password. Used when the user has forgotten their password.
+ * @returns bool - true if successfully sent the email, false otherwise
```
Comment (mahadzaryab1) [inline backend/typescript/rest/authRoutes.ts:108]: is there a `bool` being returned? I just see an empty 200 being returned.

File: frontend/src/components/auth/AuthActions.tsx [hunk 12]
```diff
@@ -17,6 +17,9 @@ const AuthActions = (): React.ReactElement => {
 
   const mode = searchParams.get("mode");
   const uid = searchParams.get("uid");
+  const isNewAccount = searchParams.get("new-account")
+    ? searchParams.get("new-account")?.toLowerCase() === "true"
+    : null;
```
Comment (mahadzaryab1) [inline frontend/src/components/auth/AuthActions.tsx:22]: is there any way we can read the parameter as a bool instead of comparing it to a string

File: frontend/src/components/auth/ForgotPassword.tsx [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+import React, { useState } from "react";
+import { useHistory } from "react-router-dom";
+
+import authAPIClient from "../../APIClients/AuthAPIClient";
+/* Images */
+import CCBCLogo from "../../images/ccbc-logo.png";
+import LoginGraphic from "../../images/Login-graphic.png";
+import ForgotPasswordEmailModal from "./ForgotPasswordEmailModal";
+
+const ForgotPassword = (): React.ReactElement => {
```
Comment (mahadzaryab1) [inline frontend/src/components/auth/ForgotPassword.tsx:25]: nit: add documentation for this component

File: frontend/src/components/auth/ForgotPassword.tsx (hunk 10 above)
Comment (mahadzaryab1) [reply frontend/src/components/auth/ForgotPassword.tsx:94]: we should probably just do a regex check to make sure the email is valid

File: frontend/src/components/auth/ForgotPasswordEmailModal.tsx [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+  ModalOverlay,
+  Text,
+} from "@chakra-ui/react";
+import React from "react";
+
+import AuthModalBackground from "../../assets/forgotPasswordModalGraphic.svg";
+
+const ForgotPasswordEmailModal = ({
+  isOpen,
+  onClose,
```
Comment (mahadzaryab1) [inline frontend/src/components/auth/ForgotPasswordEmailModal.tsx:18]: nit: create a new interface called `ForgotPasswordEmailModalProps` and put these in there. It'll allow us to document the props properly.

File: frontend/src/components/auth/ForgotPasswordEmailModal.tsx [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+  ModalContent,
+  ModalHeader,
+  ModalOverlay,
+  Text,
+} from "@chakra-ui/react";
+import React from "react";
+
+import AuthModalBackground from "../../assets/forgotPasswordModalGraphic.svg";
+
+const ForgotPasswordEmailModal = ({
```
Comment (mahadzaryab1) [inline frontend/src/components/auth/ForgotPasswordEmailModal.tsx:16]: nit: add documentation for this component

File: frontend/src/theme/index.tsx [hunk 16]
```diff
@@ -43,6 +43,19 @@ const customTheme = extendTheme({
...
     },
+    Modal: {
+      variants: {
+        forgotPasswordModal: {
+          dialog: {
+            maxWidth: "590px",
+            minWidth: "590px",
+            minHeight: "480px",
+            maxHeight: "480px",
+            borderRadius: "20px",
```
Comment (mahadzaryab1) [inline frontend/src/theme/index.tsx:54]: I'm wondering if we should just have these defined in the modal itself rather than here - that's what we seem to be doing for the other components. Thoughts?

**Review by mahadzaryab1 — COMMENTED** [review]
added a few comments

File: frontend/src/components/auth/AuthActions.tsx (hunk 12 above)
Comment (dhmi19) [reply frontend/src/components/auth/AuthActions.tsx:22]: Don't think so. I think we'll have to use a third party library which I don't think is worth it:
https://stackoverflow.com/questions/39599012/proper-way-to-parse-boolean-query-string-param-in-node-express

File: frontend/src/components/auth/ForgotPassword.tsx (hunk 13 above)
Comment (dhmi19) [reply frontend/src/components/auth/ForgotPassword.tsx:25]: What does documentation for a React component look like? I don't think we are documenting components in the project and starter code doesn't have any examples either. Could you link an example?

File: frontend/src/components/auth/ForgotPasswordEmailModal.tsx (hunk 15 above)
Comment (dhmi19) [reply frontend/src/components/auth/ForgotPasswordEmailModal.tsx:16]: Same as above regarding component documentation ^^

File: frontend/src/theme/index.tsx (hunk 16 above)
Comment (dhmi19) [reply frontend/src/theme/index.tsx:54]: Yeh that's a fair point. I really wanted to do it the Chakra UI way with defining custom styles and all but it didn't end up being as clean as I had imagined lol. I moved these inline with the component

File: frontend/src/components/auth/ForgotPassword.tsx (hunk 10 above)
Comment (dhmi19) [reply frontend/src/components/auth/ForgotPassword.tsx:94]: Added a util function to validate email addresses on the front end using regex

File: frontend/src/App.tsx (hunk 9 above)
Comment (dhmi19) [reply frontend/src/App.tsx]: deleted the inexact route

File: backend/typescript/rest/authRoutes.ts (hunk 8 above)
Comment (dhmi19) [reply backend/typescript/rest/authRoutes.ts]: done

File: backend/typescript/services/implementations/authService.ts (hunk 6 above)
Comment (dhmi19) [reply backend/typescript/services/implementations/authService.ts:187]: resolved for now as discussed

File: backend/typescript/rest/authRoutes.ts (hunk 7 above)
Comment (dhmi19) [reply backend/typescript/rest/authRoutes.ts:118]: resolved for now as discussed

File: frontend/src/components/auth/ForgotPassword.tsx (hunk 13 above)
Comment (mahadzaryab1) [reply frontend/src/components/auth/ForgotPassword.tsx:26]: https://github.com/uwblueprint/ccbc/pull/48/files

**Comment by susheelk** [comment]
Functionality seems good! 
Should the text content be different here from the create account flow?
image

**Review by susheelk — APPROVED** [review]

File: frontend/src/components/auth/ForgotPasswordEmailModal.tsx [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+import React from "react";
+
+import AuthModalBackground from "../../assets/forgotPasswordModalGraphic.svg";
+
+/**
+ * This is a model for the ForgotPasswordEmailModal component's props
+ */
+interface ForgotPasswordEmailModalProps {
+  isOpen: boolean;
+  onClose: () => void;
```
Comment (mahadzaryab1) [inline frontend/src/components/auth/ForgotPasswordEmailModal.tsx:21]: nit: document these props

File: frontend/src/components/auth/SetPassword.tsx [hunk 18]
```diff
@@ -24,12 +24,24 @@ import { AuthenticatedUser } from "../../types/AuthTypes";
 import firebaseApp from "../../utils/Firebase";
 import PasswordInputField from "../common/PasswordInputField";
 
+/**
+ * A model describing the props for SetPassword component
+ */
 export type SetPasswordProps = {
   email: string;
   uid: string;
+  isNewAccount: boolean;
```
Comment (mahadzaryab1) [inline frontend/src/components/auth/SetPassword.tsx:33]: same here

File: frontend/src/components/auth/VerifyAccessCode.tsx [hunk 19]
```diff
@@ -26,7 +26,13 @@ import firebaseApp from "../../utils/Firebase";
 import PasswordInputField from "../common/PasswordInputField";
 import { SetPasswordProps } from "./SetPassword";
 
-const VerifyAccessCode = ({ uid }: { uid: string }): React.ReactElement => {
+const VerifyAccessCode = ({
+  uid,
+  isNewAccount,
+}: {
+  uid: string;
+  isNewAccount: boolean;
```
Comment (mahadzaryab1) [inline frontend/src/components/auth/VerifyAccessCode.tsx:34]: same comment about setting these as props in an interface

**Review by mahadzaryab1 — APPROVED** [review]
minor comments - otherwise LGTM!
